src/consul.rs

```diff
@@ -91,4 +102,58 @@ impl ConsulClient {
         println!("Dropped Consul lock");
         Ok(())
     }
+
+    /// Retrieves a list of all registered services and parses their tags into DnsTag
+    pub async fn fetch_service_tags(&self) -> Result<Vec<DnsTag>, anyhow::Error> {
+        let mut services_url = self.kv_api_base_url.join("catalog/")?.join("services")?;
+
+        // Set dc if it is provided in the config
+        if let Some(dc) = &self.datacenter {
+            services_url.query_pairs_mut().append_pair("dc", dc);
+        }
+
+        // Add a filter to only match "normal" Consul services
+        services_url
+            .query_pairs_mut()
+            .append_pair("filter", "ServiceKind==\"\"");
+
+        let response = self
+            .http_client
+            .get(services_url)
+            .send()
+            .await?
+            .error_for_status()?
+            .json::<HashMap<String, Vec<String>>>()
+            .await?;
+
+        let mut dns_tags = Vec::new();
+        for (_service_name, tags) in response {
+            if let Some(dns_tag) = parse_dns_tags(tags) {
+                dns_tags.push(dns_tag);
+            }
+        }
+
+        Ok(dns_tags)
+    }
+}
+
+fn parse_dns_tags(tags: Vec<String>) -> Option<DnsTag> {
+    let mut tag_map: HashMap<String, String> = HashMap::new();
+    for tag in tags.iter().filter(|t| t.starts_with("external-dns.")) {
+        let parts: Vec<&str> = tag.split('=').collect();
+        if parts.len() == 2 {
+            tag_map.insert(parts[0].to_string(), parts[1].to_string());
+        }
+    }
```

Comment:
We need to parse any number of records, namespaced by the given identifier.

```suggestion
    const PREFIX: &'static str = "external-dns.";

    let mut dns_tags: HashMap<String, HashMap<String, String>> = HashMap::new();
    for tag in tags.into_iter() {
        let Some(rest) = tag.strip_prefix(PREFIX) else {
            continue;
        };
        let Some((identifier, rest)) = rest.split_once('.') else {
            continue;
        };
        let Some((field, value)) = rest.split_once('=') else {
            continue;
        };
        dns_tags
            .entry(identifier.to_string())
            .or_default()
            .insert(field.to_string(), value.to_string());
    }
```